src/attributes.jl

```diff
@@ -0,0 +1,232 @@
+"""
+    HDF5.Attribute
+
+A HDF5 attribute: this is a piece of metadata attached to an HDF5 `Group` or
+`Dataset`. It acts like a `Dataset`, in that it has a defined datatype and
+dataspace, and can `read` and `write` data to it.
+
+See also
+- [`open_attribute`](@ref)
+- [`create_attribute`](@ref)
+- [`read_attribute`](@ref)
+- [`write_attribute`](@ref)
+- [`delete_attribute`](@ref)
+"""
+mutable struct Attribute
+    id::API.hid_t
+    file::File
+
+    function Attribute(id, file)
+        dset = new(id, file)
+        finalizer(close, dset)
+        dset
+    end
+end
+Base.cconvert(::Type{API.hid_t}, attr::Attribute) = attr
+Base.unsafe_convert(::Type{API.hid_t}, attr::Attribute) = attr.id
+function Base.close(obj::Attribute)
+    if obj.id != -1
+        if obj.file.id != -1 && isvalid(obj)
+            API.h5a_close(obj)
+        end
+        obj.id = -1
+    end
+    nothing
+end
+name(attr::Attribute) = API.h5a_get_name(attr)
+
+
+datatype(dset::Attribute) = Datatype(API.h5a_get_type(checkvalid(dset)), file(dset))
+dataspace(attr::Attribute) = Dataspace(API.h5a_get_space(checkvalid(attr)))
+
+function Base.write(obj::Attribute, x)
+    dtype = datatype(x)
+    try
+        write_attribute(obj, dtype, x)
+    finally
+        close(dtype)
+    end
+end
+
+"""
+    read_attribute(parent::Union{File,Group,Dataset,Datatype}, name::AbstractString)
+
+Read the value of the named attribute on the parent object.
+
+# Example
+```julia-repl
+julia> HDF5.read_attribute(g, "time")
+2.45
+```
+"""
+function read_attribute(parent::Union{File,Group,Dataset,Datatype}, name::AbstractString)
```

Comment:
I'll admit that there aren't any clear rules, but there are a couple of reasons to keep it here:

- they all use the attribute (`h5a_XXX`) C API functions
- attributes are an "extra" thing stuck on top of HDF5 (i.e. you can define groups, datasets etc without referring to attributes at all)
- in this particular case, it is a method that applies to multiple objects (so if you were to put it with those, it would need to be defined multiple times)